src/app/src/main/java/com/example/bookmark/SignupActivity.java

```diff
@@ -0,0 +1,200 @@
+        signUpButton = findViewById(R.id.signup_signup_button);
+
+        userNameLayout = findViewById(R.id.signup_username_textInputLayout);
+        firstNameLayout = findViewById(R.id.signup_firstname_textInputLayout);
+        lastNameLayout = findViewById(R.id.signup_lastname_textInputLayout);
+        emailAddressLayout = findViewById(R.id.signup_email_textInputLayout);
+        phoneNumberLayout = findViewById(R.id.signup_phonenumber_textInputLayout);
+
+        ArrayList<EditText> formTexts = new ArrayList<>();
+        formTexts.add(userNameEditText);
+        formTexts.add(firstNameEditText);
+        formTexts.add(lastNameEditText);
+
+        ArrayList<TextInputLayout> formLayouts = new ArrayList<>();
+        formLayouts.add(userNameLayout);
+        formLayouts.add(firstNameLayout);
+        formLayouts.add(lastNameLayout);
+
+         signUpButton.setOnClickListener(view -> {
+            boolean invalidField = false;
+            // double check all form fields filled
+            for (int i = 0; i < formTexts.size(); i++) {
+                if (!validateEditTextEmpty(formTexts.get(i), formLayouts.get(i))) {
+                    invalidField = true;
+                }
+            }
+            // check email and phone validation
+            if (!checkIfEditTextValidEmail(emailAddressEditText, emailAddressLayout)) {
+                invalidField = true;
+            }
+            if (!checkIfEditTextValidPhoneNumber(phoneNumberEditText, phoneNumberLayout)) {
+                invalidField = true;
+            }
+            if (invalidField) {
+                return;
+            }
+
+            // check username is not in db already
+            String username = userNameEditText.getText().toString();
+            firebaseProvider.retrieveUserByUsername(username,
+                user -> {
+                    if (user == null) {
+                        // Get values from fields and create user object
+                        String firstName = firstNameEditText.getText().toString();
+                        String lastName = lastNameEditText.getText().toString();
+                        String emailAddress = emailAddressEditText.getText().toString();
+                        String phoneNumber = phoneNumberEditText.getText().toString();
+
+                        // add user and go back to main activity
+                        User newUser = new User(username, firstName, lastName, emailAddress, phoneNumber);
+                        addUser(newUser, firebaseProvider);
+                        Intent intent = new Intent(SignupActivity.this, MainActivity.class);
+                        intent.putExtra("SIGNED_UP_USERNAME", username);
+                        startActivity(intent);
+                    } else {
+                        userNameLayout.setError("Username already registered. Please choose another");
+                        return;
```

Comment:
This return statement doesn't do anything meaningful. Just commenting on it since you might have thought it would return from the `onCreate` method, when in actuality it returns from the lambda function you're passing to `firebaseProvider.retrieveUserByUsername`.

Reply:
Good catch